### PR TITLE
Changes in buildings

### PR DESCRIPTION
**Pull request by tomas2211**
Adding blocks & semantic2vef

### CODE REVIEW

File: semantic/mesh/mesh.cpp [hunk 1]
```diff
@@ -59,9 +60,24 @@ geometry::Mesh mesh(const Building &building, const MeshConfig &config
     const auto origin(origin_ + building.origin);
 
     geometry::Mesh m;
-    for (const auto &roof : building.roofs) {
-        detail::append(m, mesh(roof, config, origin + roof.center));
+    if (building.mesh.vertices.size())
```
Comment (matyas-hollmann) [inline semantic/mesh/mesh.cpp]: Check against `! .empty()` semantically signals the intention better.

File: semantic/mesh/multipolymesh.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        return res;
+    }
+    return Material::default_;
+}
+
+geometry::Mesh mesh(const geometry::MultiPolyMesh<std::string>& mpmesh,
+                    const MeshConfig& config,
+                    const math::Point3& origin)
+{
+    (void) config;
```
Comment (matyas-hollmann) [inline semantic/mesh/multipolymesh.cpp]: Isn't it better to not give parameters we are not planning to use names? E.g. 
```
geometry::Mesh mesh(const geometry::MultiPolyMesh<std::string>& mpmesh,
                    const MeshConfig& /* config */,
                    const math::Point3& origin)
```
Or in the future (C++17, use parameter `[[maybe_unused]]`).

File: semantic/mesh/multipolymesh.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    geometry::Mesh m = mpmesh.triangulateFaces(&faceLabels);
+
+    for (auto& v : m.vertices)
+    {
+        v += origin;
+    }
+
+    for (std::size_t i = 0; i < m.faces.size(); i++)
+    {
+        m.faces[i].imageId = +string2Material(faceLabels[i]);
```
Comment (matyas-hollmann) [inline semantic/mesh/multipolymesh.cpp:56]: Interesting idea to override the unary plus operator to get the underlying numerical representation. In the standard (C++23) this function will be called `std::to_underlying`. Can we rename it to something more descriptive? `toUnderlying` \ `toInteger` \ `asInteger`? Thanks.

File: semantic/tools/data/semantic.mtl [hunk 4]
```diff
@@ -20,3 +20,7 @@ Kd 0.1608 0.2863 0.1020
 # brown tree trunk
 newmtl tree-trunk
 Kd 0.3255 0.2078 0.0392
+
+# terrace color - same as facade
+newmtl terrace
+Kd 0.9137 0.6431 0.2784
```
Comment (matyas-hollmann) [inline semantic/tools/data/semantic.mtl]: Missing newline at the end of the file.

File: semantic/tools/semantic2vef.cpp [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+public:
+    Semantic2Vef()
+        : service::Cmdline("semantic2vef", BUILD_TARGET_VERSION)
+    {}
+
+private:
+    virtual void configuration(po::options_description &cmdline
+                               , po::options_description &config
+                               , po::positional_options_description &pd)
+        override;
```
Comment (matyas-hollmann) [inline semantic/tools/semantic2vef.cpp:76]: I would prefer to stick to the best practice and use only `override` keyword here.  Same for the other functions.

> CppCoreGuidelines:C.128: Virtual functions should specify exactly one of virtual, override, or final.

File: semantic/tools/semantic2vef.cpp [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+         ->default_value(false)
+         , "Generate closed surface.")
+        ("txtColorWidth", po::value(&txtColorWidth_)
+         ->default_value(txtColorWidth_)
+         , "Width of individual color squares in generated texture image.")
+        ;
+
+    pd.add("output", 1).add("input", -1);
+
+    (void) config;
```
Comment (matyas-hollmann) [inline semantic/tools/semantic2vef.cpp]: I would prefer not to use not to give parameters we are not planning to use names. The `(void)` casting hack to silence the compiler warning is not the best practice.

File: semantic/tools/semantic2vef.cpp [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    }
+    return false;
+}
+
+cv::Mat3b createTextureImg(const int colorWidth)
+{
+    auto materials = enumerationValues(semantic::Material());
+    cv::Mat3b img(colorWidth, materials.size() * colorWidth);
+
+    for (auto& m : materials)
```
Comment (matyas-hollmann) [inline semantic/tools/semantic2vef.cpp]: `const auto&` if we aren't modifying materials.

File: semantic/mesh/multipolymesh.cpp (hunk 3 above)
Comment (tomas2211) [reply semantic/mesh/multipolymesh.cpp:56]: I think this should be discussed with the author of the operator, @vaclavblazek and I am not sure the discussion should be a part of this PR.
https://github.com/melowntech/libsemantic/blob/388bc4a638aaca62ef577b44c485e59cb38014b3/semantic/mesh.hpp#L88

File: semantic/tools/semantic2vef.cpp (hunk 5 above)
Comment (tomas2211) [reply semantic/tools/semantic2vef.cpp:76]: Style copied from 
https://github.com/melowntech/libsemantic/blob/388bc4a638aaca62ef577b44c485e59cb38014b3/semantic/tools/semantic2gpkg.cpp#L31-L42
but ok, `virtual` does not have to be there.

File: semantic/mesh/multipolymesh.cpp (hunk 3 above)
Comment (vaclavblazek) [reply semantic/mesh/multipolymesh.cpp:56]: I've found it somewhere as an idiom how to covert enum to int.

Comment (zap150) [reply semantic/mesh/multipolymesh.cpp:56]: I would vote for a direct cast or a function called as Matyas suggests. I assume there's a multitude of such hacks to get the underlying type from a typed enum, this one requires a foreign user to search for the implementation of the overloaded operator. If we do not want to change old code, I suggest to use the direct cast in the new parts.

Comment (vaclavblazek) [reply semantic/mesh/multipolymesh.cpp:56]: please use what's already there -- this is not contest in changing existing code
be pedant in your OWN code, pls

File: semantic/io.cpp [hunk 8]
```diff
@@ -64,7 +64,18 @@ void parse(math::Point3 &point, const Json::Value &value)
     Json::unpack(value, "Point3", point(0), point(1), point(2));
 }
 
-void parse(math::Points3 &points, const Json::Value &value)
+void parse(std::string& str, const Json::Value &value)
```
Comment (matyas-hollmann) [inline semantic/io.cpp]: Formatting inconsistency of ampersands; I personally prefer to emphasize it's a modifier of a type and write them adjacent to the type. E.g. `void parse(std::string& str, const Json::Value& value)`.

Comment (vaclavblazek) [reply semantic/io.cpp]: And I exactly prefer it the other way around :man_shrugging: 
Technically, it modifes the variable, not the type.

Comment (ghost) [reply semantic/io.cpp]: It’s a matter of preference; but consistency is the key. It’s quite weird to combine both styles on a single line. Btw Bjarne Stroustrup, Scott Meyers, Herb Sutter, et al. write it adjacent to the type.

Comment (zap150) [reply semantic/io.cpp]: This is again a matter of taste, in fact I am used to center alignment... In our clang-format we defined left alignment. Choose whatever, I only suggest to be consistent - at least within a line.

Comment (tomas2211) [reply semantic/io.cpp]: Fixed

**Pull request merged**